statically link sunmscapi

kkriske opened the pull request:
This PR adds support to statically link `sunmscapi`, required for the `sun_security_mscapi` package.

First of all, this still has a **critical issue** that needs to be taken care of before it would be ready to merge.
Additionally, some extra JNI configuration can be added, and the reachability handler can be refined.

The critical issue with this addition, happens when both `sunmscapi` and `sunec` are added:
`sunec.lib(ECC_JNI.obj) : error LNK2005: ThrowException already defined in sunmscapi.lib(security.obj)`

The offending definitions:
https://github.com/graalvm/labs-openjdk-11/blob/0cbcc0176b3404262b95c180ea716e9b0cdeccf6/src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp#L132
https://github.com/graalvm/labs-openjdk-11/blob/0cbcc0176b3404262b95c180ea716e9b0cdeccf6/src/jdk.crypto.ec/share/native/libsunec/ECC_JNI.cpp#L50

I can handle the refinement of the reachability hander and the JNI config, but I need some help for the linking error.

pejovica commented:
> The critical issue with this addition, happens when both `sunmscapi` and `sunec` are added:
`sunec.lib(ECC_JNI.obj) : error LNK2005: ThrowException already defined in sunmscapi.lib(security.obj)`

Unfortunately, this would have to be fixed in the JDK. However, `sunec` has been removed in JDK 16 (see [JDK-8235710](https://bugs.openjdk.java.net/browse/JDK-8235710)), so you could adjust the PR so that static linking is used only on JDK 16+, while on older JDKs we would continue to use dynamic linking (which will now work properly, see 5a34e716b111961e5a40e4c641ca3ba318d01ce5).

kkriske commented:
I added a bit more JNI stuff, not sure if that's desired or not. And I added the suggested check for jdk 16. I have yet to try it out locally as I can't seem to build a java 16 version of GraalVM locally.

My build fails on the generation of polybench.exe, which I don't think is required so I'm probalby building it wrong? What would be the correct mx command to get a simple native-image capable local build?

kkriske commented:
Managed to get the build working and was able to try it out. It should be ready for review now.

pejovica commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SecurityServicesFeature.java:
```diff
@@ -384,6 +405,37 @@ private static void linkSunEC(DuringAnalysisAccess a) {
+    }
+
+    private static void registerLoadKeysOrCertificateChains(DuringAnalysisAccess a) {
+        JNIRuntimeAccess.register(clazz(a, "java.util.ArrayList"));
```
Not necessary, because the constructor/method registration will also register the class.
```suggestion
```

kkriske replied:
done

pejovica commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SecurityServicesFeature.java:
```diff
@@ -384,6 +405,37 @@ private static void linkSunEC(DuringAnalysisAccess a) {
+    }
+
+    private static void registerGenerateCKeyPair(DuringAnalysisAccess a) {
+        JNIRuntimeAccess.register(clazz(a, "sun.security.mscapi.CKeyPair"));
```
```suggestion
```

kkriske replied:
done

pejovica commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SecurityServicesFeature.java:
```diff
@@ -384,6 +405,37 @@ private static void linkSunEC(DuringAnalysisAccess a) {
+    }
+
+    private static void registerCPrivateKeyOf(DuringAnalysisAccess a) {
+        JNIRuntimeAccess.register(clazz(a, "sun.security.mscapi.CPrivateKey"));
```
```suggestion
```

kkriske replied:
done

pejovica commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SecurityServicesFeature.java:
```diff
@@ -309,6 +310,26 @@ public void beforeAnalysis(BeforeAnalysisAccess access) {
+
+        if (isWindows()) {
+            // JNI config required by sunmscapi
+            access.registerReachabilityHandler(SecurityServicesFeature::registerLoadKeysOrCertificateChains,
+                            method(access, "sun.security.mscapi.CKeyStore", "loadKeysOrCertificateChains", String.class));
+            access.registerReachabilityHandler(SecurityServicesFeature::registerGenerateCKeyPair,
+                            method(access, "sun.security.mscapi.CKeyPairGenerator$RSA", "generateCKeyPair", String.class, int.class, String.class));
+            access.registerReachabilityHandler(SecurityServicesFeature::registerCPrivateKeyOf,
+                            method(access, "sun.security.mscapi.CKeyStore", "storePrivateKey", String.class, byte[].class, String.class, int.class),
+                            method(access, "sun.security.mscapi.CSignature", "importECPublicKey", String.class, byte[].class, int.class),
+                            method(access, "sun.security.mscapi.CSignature", "importPublicKey", String.class, byte[].class, int.class));
```
Could you factor this out into a helper method, perhaps `registerSunMSCAPIReachabilityHandlers`.

pejovica reviewed (commented):
> I added a bit more JNI stuff, not sure if that's desired or not.

It's definitely desirable. 😁

Is this the entire JNI config? If not, would you be able to provide the rest?

kkriske commented:
> Is this the entire JNI config? If not, would you be able to provide the rest?

I _think_ this is everything based on the source code I found [here](https://github.com/graalvm/labs-openjdk-11/blob/master/src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp). The jdk 16 code seems to be identical as well.

The only thing that may be missing is [some exception classes](https://github.com/graalvm/labs-openjdk-11/blob/60e0ae5d1ada661540ca16388b3315580ebecc78/src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp#L52-L60). But I'm not sure what to do with these.
Do I just put them in one entry like [here](https://github.com/oracle/graal/blob/f07d5e3b858b2441e942958d8f329238ff0b0c87/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationJava.java#L62-L69)?

pejovica commented:
> > Is this the entire JNI config? If not, would you be able to provide the rest?
> 
> I _think_ this is everything based on the source code I found [here](https://github.com/graalvm/labs-openjdk-11/blob/master/src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp). The jdk 16 code seems to be identical as well.

Great, then we shouldn't need an agent to collect the `sunmscapi` config anymore.

> The only thing that may be missing is [some exception classes](https://github.com/graalvm/labs-openjdk-11/blob/60e0ae5d1ada661540ca16388b3315580ebecc78/src/jdk.crypto.mscapi/windows/native/libsunmscapi/security.cpp#L52-L60). But I'm not sure what to do with these.
> Do I just put them in one entry like [here](https://github.com/oracle/graal/blob/f07d5e3b858b2441e942958d8f329238ff0b0c87/substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationJava.java#L62-L69)?

I think we should register them here in a reachability handler, say `registerSunMSCAPIConfig`, which is triggered when the class `sun.security.mscapi.SunMSCAPI` becomes reachable. In fact, you can even move the registration of other handlers (suggested in https://github.com/oracle/graal/pull/3300#discussion_r662244696) into this new handler instead of the helper method.

kkriske commented:
The config is now located in a `registerSunMSCAPIConfig` which is registered as a reachabillity handler.
I duplicated the reachabillity handler for the linking, which could have 2 remarks:
* It duplicates the reachabillity handler, but I think it's clearer if the JNI config and the library linking are in separate methods
* I'm not sure if `registerClassInitializerReachabilityHandler` is right, or if it should be a normal class reachabillity handler.

pejovica commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SecurityServicesFeature.java:
```diff
@@ -309,6 +310,17 @@ public void beforeAnalysis(BeforeAnalysisAccess access) {
         }
+
+        if (isWindows()) {
+            access.registerClassInitializerReachabilityHandler(SecurityServicesFeature::registerSunMSCAPIConfig, clazz(access, "sun.security.mscapi.SunMSCAPI"));
```
This should be a normal class reachability handler because the `sun.security.mscapi.SunMSCAPI` class is initialized at build time.
```suggestion
            access.registerReachabilityHandler(SecurityServicesFeature::registerSunMSCAPIConfig, clazz(access, "sun.security.mscapi.SunMSCAPI"));
```

pejovica commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SecurityServicesFeature.java:
```diff
@@ -309,6 +310,17 @@ public void beforeAnalysis(BeforeAnalysisAccess access) {
+            // statically linking sunmscapi conflicts with sunEC
+            // after the removal of sunEC in jdk 16, we can statically link sunmscapi
+            if (JavaVersionUtil.JAVA_SPEC >= 16) {
+                access.registerClassInitializerReachabilityHandler(SecurityServicesFeature::linkSunMSCAPI, clazz(access, "sun.security.mscapi.SunMSCAPI"));
```
Same here.
```suggestion
                access.registerReachabilityHandler(SecurityServicesFeature::linkSunMSCAPI, clazz(access, "sun.security.mscapi.SunMSCAPI"));
```

pejovica reviewed (commented):
> I duplicated the reachabillity handler for the linking...

This is perfectly fine.

kkriske commented:
@pejovica changes applied.

pejovica reviewed (commented):
Two more corrections are needed, but other than that, it looks good!

kkriske commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SecurityServicesFeature.java:
```diff
@@ -384,6 +396,49 @@ private static void linkSunEC(DuringAnalysisAccess a) {
+    private static void registerLoadKeysOrCertificateChains(DuringAnalysisAccess a) {
+        JNIRuntimeAccess.register(constructor(a, "java.util.ArrayList"));
+        JNIRuntimeAccess.register(clazz(a, "sun.security.mscapi.CKeyStore$MY"));
+        JNIRuntimeAccess.register(clazz(a, "sun.security.mscapi.CKeyStore$ROOT"));
```
Interesting, I did not know that.

kkriske commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/SecurityServicesFeature.java:
```diff
@@ -384,6 +396,49 @@ private static void linkSunEC(DuringAnalysisAccess a) {
+                        method(a, "sun.security.mscapi.CKeyPairGenerator$RSA", "generateCKeyPair", String.class, int.class, String.class));
+        a.registerReachabilityHandler(SecurityServicesFeature::registerCPrivateKeyOf,
+                        method(a, "sun.security.mscapi.CKeyStore", "storePrivateKey", String.class, byte[].class, String.class, int.class),
+                        method(a, "sun.security.mscapi.CSignature", "importECPublicKey", String.class, byte[].class, int.class),
+                        method(a, "sun.security.mscapi.CSignature", "importPublicKey", String.class, byte[].class, int.class));
```
fixed

kkriske commented:
requested changes applied.

pejovica reviewed (approved):
Excellent! I'll start the integration.

mttcr commented:
Thank you for the feature, it was awaited for #1410 ! 
I'm not sure this is the right place to ask, but could something similar be done to statically link j2pkcs11 under linux? this would solve #2552

sgeeroms commented:
Hi, we have an issue the when no java is installed the test reverts to dynamic linking, but then it fails to link them off course. Is there a way to force static linking in case there is no java installed. Or do I maybe not really understand what static linking beholds?
See the error we get on a machine without java installed. It works fine on a pc with java 17 installed (and in the PATH).
![image](https://github.com/user-attachments/assets/eb62c25e-35f5-47aa-ba63-52ec16295420)
